src/coreComponents/physicsSolvers/fluidFlow/kernels/immiscibleMultiphase/ImmiscibleMultiphaseKernels.hpp

```diff
@@ -0,0 +1,1420 @@
+   * @tparam FUNC the type of the function that can be used to customize the computation of the flux
+   * @param[in] iconn the connection index
+   * @param[inout] stack the stack variables
+   * @param[in] NoOpFunc the function used to customize the computation of the flux
+   */
+  template< typename FUNC = NoOpFunc >   // should change to multiphase
+  GEOS_HOST_DEVICE
+  void computeFlux( localIndex const iconn,
+                    StackVariables & stack,
+                    FUNC && kernelOp = NoOpFunc{} ) const
+  {
+    // first, compute the transmissibilities at this face                                             // get k and dk/dP from global arrays
+    // and place in stack
+    m_stencilWrapper.computeWeights( iconn,
+                                     m_permeability,
+                                     m_dPerm_dPres,
+                                     stack.transmissibility,
+                                     stack.dTrans_dPres );
+
+    localIndex k[2];
+    localIndex connectionIndex = 0;
+
+    for( k[0] = 0; k[0] < stack.numFluxElems; ++k[0] )
+    {
+      for( k[1] = k[0] + 1; k[1] < stack.numFluxElems; ++k[1] )
+      {
+        // clear working arrays
+        real64 densMean[numEqn]{};
+        real64 dDensMean_dP[numEqn][2]{};
+
+        real64 presGrad[numEqn]{};
+        real64 dPresGrad_dP[numEqn][2]{};
+
+        real64 gravHead[numEqn]{};
+        real64 dGravHead_dP[numEqn][2]{};
+
+        real64 capGrad[numEqn]{};
+        real64 dCapGrad_dP[numEqn][2]{};
+        real64 dCapGrad_dS[numEqn][2]{};
+
+        real64 fluxVal[numEqn]{};
+        real64 dFlux_dP[numEqn][2]{};
+        real64 dFlux_dS[numEqn][2]{};
+
+        real64 mobility[numEqn]{};
+        real64 dMob_dP[numEqn][2]{};
+        real64 dMob_dS[numEqn][2]{};
+
+        real64 const trans[2] = { stack.transmissibility[connectionIndex][0], stack.transmissibility[connectionIndex][1] };
+        real64 const dTrans_dP[2] = { stack.dTrans_dPres[connectionIndex][0], stack.dTrans_dPres[connectionIndex][1] };
+
+        // cell indices
+        localIndex const seri[2]  = {m_seri( iconn, k[0] ), m_seri( iconn, k[1] )};
+        localIndex const sesri[2] = {m_sesri( iconn, k[0] ), m_sesri( iconn, k[1] )};
+        localIndex const sei[2]   = {m_sei( iconn, k[0] ), m_sei( iconn, k[1] )};
+
+        // loop over phases
+        for( integer ip = 0; ip < m_numPhases; ++ip )
+        {
+          // calculate quantities on primary connected cells
+          for( integer ke = 0; ke < 2; ++ke )
+          {
+            // density
+            real64 const density  = m_dens[seri[ke]][sesri[ke]][sei[ke]][0][ip];         // r = rho1 || rho2
+            real64 const dDens_dP = m_dDens_dPres[seri[ke]][sesri[ke]][sei[ke]][0][ip][Deriv::dP];  // dr/dP = dr1/dP1 || dr2/dP
+
+            // average density and derivatives
+            densMean[ip] += 0.5 * density;          // rho = (rho1 + rho2) / 2
```

Comment:
this treatment was recently modified, see https://github.com/GEOS-DEV/GEOS/pull/3337

Reply:
New treatment implemented in kernel:
 `void computeFlux( localIndex const iconn, StackVariables & stack, FUNC && kernelOp = NoOpFunc{} ) const`

Reply:
sorry we had to revert that new treatment and make it an option, see https://github.com/GEOS-DEV/GEOS/pull/3467
somehow old treatment works much better for spe11 and it is not a good time to spoil it

Reply:
We adapted the code to make it optional in the immiscible solver as well, using the parameter `m_checkPhasePresenceInGravity`